keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,449 @@
+
+### Risks and Mitigations
+
+A poor-implemented queue may not function well, in both functionality and performance.
+But the default scheduler works as before and thus won't be impacted.
+
+## Design Details
+
+The current internal queue is coupled with `cache`, `Option` and other implementations
+(e.g., the queue is initialized with specified parameters like `LessFunc`, `SharedInformerFactory`,
+and `Option` by [factory.go#create](https://github.com/kubernetes/kubernetes/blob/f1f0183d2bbcde33024b2a05d6f39df32f11e037/pkg/scheduler/factory.go#L172)),
+this solution will not decouple them as this has no impact on the goal of custom queue,
+and such logics can be enhanced in the future on demand.
+
+Some of the items like `SchedulingQueue`, `Option` that are only used by the internal
+queue now will be exposed at [interface.go](https://github.com/kubernetes/kubernetes/blob/fb9cafc99be94a73d9b92545164dbf336bbd230a/pkg/scheduler/framework/interface.go),
+users of custom queue can access them.
+
+To implement a custom queue, what an user needs to do is to copy the codes of the current
+internal queue (e.g., scheduling_queue.go), update some logics to get the custom queue,
+and then pass the custom queue to the scheduler like the plugins. With this way, user's
+logics will override the logics of the current internal queue.
+
+The codes will be updated as following:
+
+1. Make the interface `SchedulingQueue` as a public one, so that users can implement it
+for the custom queue.
+
+2. Provide a new function `WithCustomQueue`, users can register the custom queue with this
+function like other plugins.
+
+    ```go
+         command := app.NewSchedulerCommand(
+             app.WithPlugin(coscheduling.Name, coscheduling.New),
+             ......
+       +     app.WithCustomQueue(customQueue.New),
+         )
+    ```
+
+3. The registered custom queue will be recorded by updated `Registry` struct.
+
+    ```go
+      type Registry struct {
+          Pf          map[string]PluginFactory
+          CustomQueue schedulingQueue.SchedulingQueue
+      }
+    ```
+
+4. In function `(c *Configurator) create()`, custom queue will be passed to the scheduler
+if it exists, or the current internal queue will be used.
+
+    ```go
+      - podQueue := internalqueue.NewSchedulingQueue(
+      -         lessFn,
+      +
+      + var podQueue framework.SchedulingQueue
+      + if c.registry.CustomQueue != nil {
+      +         podQueue = c.registry.CustomQueue
+      + } else {
+      +         podQueue = &internalqueue.PriorityQueue{}
+      + }
+      + podQueue.Init(...)
+    ```
+
+5. Basic inputs for queue's initialization will not be changed, a new method `Init` is added in
+the interface `SchedulingQueue`, users must implement it.
+
+    ```go
+      + func (p *PriorityQueue) Init(
+      +        lessFn framework.LessFunc,
+      +        informerFactory informers.SharedInformerFactory,
+      +        opts ...framework.Option,
+      + ) {
+    ```
+
```

Comment:
Alternatively, you can ass a stanza about the implementation practice/suggestion here:

```
To get started with the implementation, an easy way is to copy the codes of the current
internal queue (i.e., scheduling_queue.go), and then update the logic in-place to adapt to
your business needs  (it's very like how some out-of-tree PostFilter plugin refers to the defaultpreemption plugin).
With the custom queue handle, you can pass it to the scheduler like the plugins.
Following this way, the user's custom scheduler queue will override the upstream internal queue.
```

Reply:
added.